### DISCUSSION TITLE
True 30FPS?

### DISCUSSION THREAD

**Discussion by vibes24** [gopro/labs#437] [Feature Requests]
I noticed that Labs can give you true 24fps. I would like to request true 30FPS capabilities to be added to GoPro Labs.
Often I mix shots from my phone and gopro. I noticed that modern cameras now shoot in 30fps for digital tv/screens. I would like my gopro footage to be in 30fps too to match my other shots. Moreover, various apps apply effects on 30fps. I lost 2 frames when editing, so there's mismatch.

I look forward to true 30 FPS, thank you.

**Comment by dnewman-gpsw** [comment]
Please list camera models that have accurate 30.0fps, other than cellphones, which only approximates 30/29.97 (they aren't very accurate.)  29.97fps is the standard.  

Here is a list of popular models that don't:
FujiFilm XT3/4/5 - no 30.0, only 29.97, 59.94, 50.0, 23.976 and 24.0 (the standards.) 
Panasonic Lumix GH5/6 - no 30.0 see https://www.panasonic.com/sg/consumer/camera-camcorder/lumix-g-series/lumix-g-mirrorless-camera/dc-gh5.specs.html
Canon EOS R5 - no 30.0 see https://www.canon.com.au/cameras/-/media/files/canon/product-brochure/cameras/mirrorless-cameras/eos-r5-tech-sheet.ashx
Nikon R6ii - no 30.0, has this great note: Actual frame rates for 120p, 100p, 60p, 50p, 30p, 25p, and 24p are 119.88, 100, 59.94, 50, 29.97, 25, and 23.976 fps respectively

**Comment by vibes24** [comment]
Thanks for your detailed response. My Pixel 7 Pro phone for example shoots in 30FPS. I don't have a whole list of cameras with me. I am just saying that I've noticed that 30fps is emerging and that 29.97 is outdated for NTSC. I have no need for 24fps as I don't produce cinematic videos.
So you do not think it's necessary to have true 30FPS (non drop frames)? That's disappointing. Often I end up with mismatched FPS and forced me to conform, etc. I would rather them to work out of the box.

PS: Somebody mentioned in reddit that 29.97 messup After Effects. I lost 2 frames when working with 29.97 and 30 fps in Davinci Resolve. Nowadays I am working with AI tools and I am encountering issues with drop frames. I would like true 30FPS. Googling shows people asking for them with their own reasons, I've listed mine.

Apparently Blackmagic cameras offer this option.

**Comment by dnewman-gpsw** [comment]
Sorry, 30.0 is a pet annoyance of mine.  29.97 is correct, and yes, it is annoying and weird, and has a long history, but the correct thing to do. 24.0 was added from a request from Disney (info here https://gopro.com/it/it/news/gopro-labs-updates-2022.)

Adding 24.0 was not easy, 30.0 would have to same difficulties.

Technically you should conform your cellphone footage to 29.97, not your GoPro to match your cellphone.

**Reply by vibes24** [reply]
I work with AI tools which takes apart frames and rebuild them. I have lots of issues with 29.97 missing frames.

Your link have a highlight in big and bold text " You want it? GoPro Labs will create it". Looks like you are saying only if it pleases you and big people. No offense, just that I suggested a feature for my needs and I got shot down within a minute because its hard or does not meet your needs or big people's needs.

I am not saying you have to make it happen ASAP. Again, a feature request which I would like to see manifest some day. Thank you.

EDIT: PS - googling shows other people having the same needs for their own reasons.

**Reply by vibes24** [reply]
I will try conform 30 fps to 29.97 in the future. Just would like everything to work out of the box. Thanks for the suggestion.

**Reply by dnewman-gpsw** [reply]
I was simply asking for a clear reason, you gave "modern cameras now shoot in 30fps", which was not true for most cameras, whereas many modern cameras do add 24.0.  I do run a lot of Labs development, and the work has to be for productive reasons, adding one feature is often at the expense of another. 30.0 has some merit, but also has simple workarounds. The 24.0 example Disney gave does not have an easy workaround. It is not the big people, it is the concrete ideas I'm looking for.  Just make a good argument, and skip the personal attacks.

**Reply by oxine** [reply]
@dnewman-gpsw I do need true 30/60 fps in my workflow. I'm working for a large game company, belonging to the motion capture team. The  [Vicon motion capture system](https://www.vicon.com/hardware/cameras/vanguard/) runs in accurate 120fps. The mocap data generated by the system is 120fps( in *.fbx file format), then our animatiors import the motion data into animators DCCs(AutoDesk Maya, 3ds Max Motionbuilder, etc) to deal with the animation and finally export to an 30/60fps animation clip file for different project team or the AI team. We want to use a GoPro as a portable motion capture setup to record motion capture videos for animators' reference，and I believe all the animators work with motion capture system encounter the same problem: the mocap data can't be 29.97fps,  the animation clip we produced can also not be 29.97/59.94 fps, but the reference video we captured is always 29.97fps, so the reference video and the animation pose at the same frame just woundn't fit. it's realy annoying when producing high-level animation, and we do appreaciated a lot if GoPro could offer a true 30/60fps shutter rate (150fps in pal is also workable), since as you listed, lots of cameras do not support it.

**Reply by dnewman-gpsw** [reply]
This is a reason I understand. Thank you. It is being investigated for a HERO13 update.

**Reply by oxine** [reply]
> This is a reason I understand. Thank you. It is being investigated for a HERO13 update.

@dnewman-gpsw That’s great to hear! Thanks a lot for considering this—We really appreciate it.

**Reply by dnewman-gpsw** [reply]
@oxine Feature is in today's public Labs beta.

**Reply by oxine** [reply]
@dnewman-gpsw  That's so rapid! Our hardware team is currently working on shutter testing. Thanks for your great work once again!

**Comment by dnewman-gpsw** [comment]
With 30.0 sources, just interpret as 29.97, and slow the audio to 99.9% to preserve sync. No frames are lost.

**Reply by R24UD** [reply]
@dnewman-gpsw , I conclude from your response to @vibes24's request that support for 30 FPS (integer frame rate) is currently not planned and that there is a simple workaround for anyone having footage in different frame rates.
I would like to playback synchronized videos of both phones/other proprietary recording hardware (only supporting integer frame rate 30) and GoPros (Hero 11 - only supporting 29.97 FPS).
As far as I understand, you suggest to adjust the interpretation of the 30 FPS sources and slowing the playback of audio down. With only limited experience in video editing, it is not really clear for me how to translate your instructions in single, precise steps.
Could you please elaborate what exactly I would have to do to be able to playback an mp4 (30 FPS; file includes audio) from a phone in parallel with an mp4 (29.97 FPS; file includes audio) from a GoPro? Can I perform these steps with ffmpeg or which tool should I use to prepare the videos?
My current workflow is to synchronize both videos by audio in one tool (returns a frame number offset for each video) and playback and evaluate the videos in another (can only enter video offsets as time offsets). With the different frame rates, I'm clueless how to do this.
One further question regarding the 29.97 frame rate implemented for the GoPro: Does the generated file use drop frame or non-drop frame time code? Do different time codes require any special handling for my use case? Thanks for clarifying!

**Reply by dnewman-gpsw** [reply]
As your need doesn't seem to be editing related, I might be lacking in knowledge on how to help.  Editing tools do not have an issue, with 29.97 and 30fps sources.  GoPro Quik can mix phone and GoPro source without issue.  Using FFMPEG to "fix" seems to be the wrong approach. 

If you want to extract frames from the same time, this is the math.

frame = offset_in_seconds * 30.0;  // for many cellphones (not all, because cellphones can be bad at this.)
frame = offset_in_seconds * 30000 / 1001; // For all other cameras.
audio_sample =  offset_in_seconds * 48000;  // same for both

The frame's temporal drift at worst is 33ms, but using the above math sync is preserved. Whereas using 30 fps math for both source will drift further and further apart.

If you need the cameras to be precisely in sync, you shouldn't be using cellphones.

Timecode: GoPro stores timecode in frames since midnight, using non-drop timecode. This is a standard Quicktime timecode track, all timecode enabled tools should be able to read it.

**Reply by R24UD** [reply]
Thanks @dnewman-gpsw for your reply! It already helped me to improve my understanding of the situation. I think what confused me a little is [this article](https://community.gopro.com/s/article/Drop-Frame-Timecode-Information?language=en_US) stating that GoPro cameras "_support_ drop frame timecode" which sounds like it could be turned on or off. Do I misunderstand that information?
My understanding is that the timecode can be used for synchronization of several media. Furthermore, I assume that the timecode is what is displayed by media players (I will call it _video time_ below) as well as video editors and may deviate from real time under certain circumstances (see below). Please correct me if I am wrong.

Regarding the drift: The tool I use to evaluate the videos side-by-side retrieves frames based on timestamps. I read [here ](https://www.3playmedia.com/blog/drop-frame-non-drop-frame-affects-captions-subtitles/) that video time for non-drop timecode does not reflect "real time", that is, footage of one hour real time for the camera examples mentioned above will result in following video times:
```
GoPro camera (29.97 FPS NDF): 00:59:56:12
phone camera (30.00 FPS NDF): 01:00:00:00
```
So if the tool reads each frame based on the timecode, there will be an increasing drift during playback. This is what you mentioned when referring to 30 fps math for both sources, right?

Given the equations, if the synchronization tool returns frame number offsets, I could derive the time offset this way:
```
rt_offset_in_seconds = frame / 30.0 // cellphone 30.0 FPS (real time offset)
rt_offset_in_seconds = frame * 1001 / 30000 // other cameras, e.g., GoPro (real time offset)
vt_offset_in_seconds = frame / 30.0 // video time timecode offset for all cameras running at 29.97 / 30.0 FPS
```
For instance:
```
GoPro camera 1: frame 0 -> 0.000s real time and 0.000s video time
GoPro camera 2: frame 100 -> 100*1001/30000 = 3.337s real time and 100/30 = 3.333s video time
phone camera 1: frame 0 -> 0.000s real time and 0.000s video time
phone camera 2: frame 100 -> 100/30 = 3.333s real time and 100/30 = 3.333s video time
```

That is why I thought of using a tool like ffmpeg for a "fix" (e.g., compensate for the timecode difference by [changing the frame rate of the video](https://video.stackexchange.com/questions/34988/converting-video-from-30-fps-to-29-97fps-so-audio-syncs-to-other-29-97fps-videos)). Could you explain why this would not be a good solution? What else can I do given that the tool reads frames by video time?

I totally agree: Phone cameras may not be the best devices when aiming for high precision. I can live with a drift of 33 or 50 ms, the only thing I need to avoid is that the videos progressively go out of sync. For 20-30 minute videos, I would expect such an effect to be clearly visible.


I know that my questions go beyond the original question of whether the integer frame rate is supported. However, it would substantially help me to understand the implications of the GoPro's capabilities and what I can do about it. Thank you for your help!

**Reply by R24UD** [reply]
Hi @dnewman-gpsw , I know it is not your primary role to explain implications of frame rate & timecode for synchronization of different recordings, but I would really appreciate to see your view on my previous post.
I went through the official GoPro support process (hotline & chat) before posting here. The support team seemed not too involved in technical details and I as layperson as well failed to make clear that I'm not worried by frame drops in recordings but need information regarding frame rate and implications of (non) drop frame timecode. Your first response already helped me to improve my understanding, that is why I hope you can help me get rid of my confusion regarding that topic.

Sorry for bothering you again!

**Reply by dnewman-gpsw** [reply]
That linked 2017 article is wrong is so many ways, please ignore it.  

```
GoPro camera 1: frame 0 -> 0.000s real time and 0.000s video time
GoPro camera 2: frame 100 -> 100*1001/30000 = 3.337s real time and 100/30 = 3.333s video time
phone camera 1: frame 0 -> 0.000s real time and 0.000s video time
phone camera 2: frame 100 -> 100/30 = 3.333s real time and 100/30 = 3.333s video time
```
Your error here is inventing the concept of "video time", and suggestion that 100 frame in at 30fps might be the same as 100 frames in at 29.97.  They are not.  There is only real-time.  If you made a crude sync file copier, at 1000 frames 29.97 source will match 1001 frames into 30fps source.

29.97 calc: 1000*1001/30000 = 33.36666 real time
30.00 calc:  1001/30 = 33.36666 real time

So if dropped (removed) every 1001th frame from the 30.0 video, then preset at either 29.97 or 30, the files remain in sync.  This is what automatically happens when 30.0 cell phone files are added to a normal 29.97 timeline within tools like Premiere Pro or DaVinci Resolve.

You can do this with ffmpeg:
ffmpeg -i <cell_phone_input.mp4> -filter:v fps=29.97 <output.mp4>

**Comment by QRabbani** [comment]
> Sorry, 30.0 is a pet annoyance of mine. 29.97 is correct, and yes, it is annoying and weird, and has a long history, but the correct thing to do. 24.0 was added from a request from Disney (info here https://gopro.com/it/it/news/gopro-labs-updates-2022.)
> 
> Adding 24.0 was not easy, 30.0 would have to same difficulties.
> 
> Technically you should conform your cellphone footage to 29.97, not your GoPro to match your cellphone.

I'd like to second @vibes24 that allowing true non-fractional frame rates beyond 24 FPS would be useful. I do ML as well, mostly working with biosignals from human subjects speaking, which need to be synchronized on the fly with audio and video to provide continuous feedback. Fractional sample rates don't align nicely with the rates that most other signals are collected (i.e., 48 kHz audio, 1 kHz EEG, etc.).

@dnewman-gpsw Just curious, what are the technical reasons, if you can explain, that made 24 FPS difficult? I'm probably quite naive on this topic, as I'm new to video, but I presume there's a sample clock/timer under the hood that triggers the camera shutter at whatever rate is specified by the software? I would have thought it would just be a matter of exposing a new rate setting to us, the users. Are there hardware parameter adjustments/considerations that need to be made to get a new rate working well?

**Reply by dnewman-gpsw** [reply]
@QRabbani The 24HZ hack is the only hack that technically adds new modes to the camera, as the camera now has both 23.976 and 24.0 frame rate support (while the menus still only show one, there are two paths, the 24HZ option selects between them.) Labs could have made a 24.0 only version, but that would have broken a rule, Labs doesn't remove any standard features. Video frame rate, and all the clock timing, is really core, and can't easily changed on the fly. All the frame rates are created at compile time. Adding 24.0 creates about 10 new shooting modes (a fair amount of work), as each resolution and aspect ratio has its own setup. So while making a 30.0 only camera is straight forward (hard coding 29.97 to be 30.0,) making it switchable is not, and having multiple Labs versions is not fair to the QA team (and a bad user experience.) Adding 60.0 and 120.0, in addition to 30.0, is about 30 additional video modes to test and support. 

Syncing with 29.97 is not new (70 years old), while I understand 30.0 is easier for audio (1600 audio samples per frame, vs 1601.6.), you still get fractional EEG samples per frame. This is typical, systems with multiple sample rates rarely divide nicely. I gather you want 30.0, rather than 25.0 (integer EEG samples per frame) or 24.0, due to lighting under 60Hz power.  

What I'm looking for is a technical reason why 30.0 is really a game changer (over 29.97,) when all the other camera companies (except Black Magic) don't offer this mode. It is totally feasible, just looking for the argument I would need to present to warrant the engineering effort. Also any support is unlikely to get back ported, if it was implemented, it will be only on the current model camera.

I do expect the 30.0 will appear more in the future, as the reasons that drove 29.97 are long gone, we are simply dealing with compatibility with a legacy standard. Cellphone developers not complying with the video standards is the start of this change.

**Reply by QRabbani** [reply]
That's a great point about the extra QA needed to support the new setting permutations, and I totally understand the need to weigh the benefit with the effort.

That's also a great point that having fractional samples per frame is a typical problem with video and other systems. It's not really a strong argument, but simple fractional ratios like 2.5 are certainly easier to deal with in signal processing, especially on the fly.

I do have the latest GoPro Hero 11, which I've been playing around with. In my case, I'm actually most interested in 120 or 240 FPS personally, because sub-10-ms precision in my pschoacoustic/psychovisual experiments is crucial. We actually have a higher resolution 30 kHz EEG stream, which aligns perfectly with true camera rates. Having the form factor of a GoPro with this capability would definitely be a game changer in my field, for psych-adjacent researchers such as myself.

**Reply by dnewman-gpsw** [reply]
I forgot about 240.0 fps (so couple of extra new modes). At that frame rates, you will have powerline issues, unless you careful with your lighting, and if that is the case, you can shoot 200.0fps.  Have you consider that frame rate?  Just switch the camera to Anti-Flicker at 50Hz, or use QR Code p200, as all the PAL rates are integer.

**Reply by QRabbani** [reply]
Good point, I was actually thinking about this originally. Are the PAL rates exact then?

I'm in the US, and the one thing I noticed that steered me away from this option was that I started getting weird light flickering in my scene when I switched the anti-flicker mode to 50 Hz, and I assumed it was because there's actually notable 60 Hz flickering (at least in that one room I tried it in).

**Reply by dnewman-gpsw** [reply]
All PAL rates are integer, 25,50,100 & 200.  You will just need to have upgrade your lighting, as most consumer LED and incandescent will flicker at 200 or 240fps.

**Reply by QRabbani** [reply]
Alright, I'll look into it. Thanks for your prompt replies!
